src/host-config/redsocks.ts

```diff
@@ -0,0 +1,352 @@
+					if (Object.keys(parsed).length === 7) {
+						conf.redsocks = parsed as unknown as ReadableProxyConfig;
+					}
+				}
+			}
+		}
+
+		return conf;
+	}
+
+	public static patch(
+		currentConf: RedsocksConfig,
+		inputConf: Partial<{
+			base: BaseConfig;
+			redsocks: Partial<WriteableProxyConfig>;
+		}>,
+	): RedsocksConfig {
+		const patchedConf: RedsocksConfig = {};
+
+		// If input is empty, redsocks config should be removed
+		if (!inputConf || Object.keys(inputConf).length === 0) {
+			return patchedConf;
+		}
+
+		if (
+			inputConf.redsocks &&
+			typeof inputConf.redsocks === 'object' &&
+			Object.keys(inputConf.redsocks).length > 0
+		) {
+			// Only include base config if redsocks config is present
+			patchedConf.base = {
+				...defaultBaseConfig,
+			};
+			// This method assumes that currentConf is a full ProxyConfig
+			// for backwards compatibility, so patchedConf.redsocks should
+			// also be a full ProxyConfig.
+			patchedConf.redsocks = {
+				...currentConf.redsocks,
+				...inputConf.redsocks,
+				// local_ip & local_port are not configurable
+				local_ip: '127.0.0.1',
+				local_port: 12345,
+			} as ReadableProxyConfig;
+		}
+
+		return patchedConf;
+	}
+
+	private static stringifyBlock(
+		label: string,
+		block: BaseConfig | ReadableProxyConfig,
+	): string {
+		const lines = Object.entries(block).map(
+			([key, value]) => `\t${key} = ${value};`,
+		);
+		return `${label} {\n${lines.join('\n')}\n}`;
+	}
+
+	private static parseBlock(
+		block: string,
+		label: string,
+	): Record<string, string> {
+		const parsedBlock: Record<string, string> = {};
+		// Skip first and last line (`$BLOCK_LABEL {` and `}`)
+		const lines = block.split('\n').slice(1, -1);
+		for (const line of lines) {
+			let [key, value] = line.split('=').map((s) => s.trim());
+
+			// Remove trailing semicolon
+			if (value) {
+				if (value.endsWith(';')) {
+					value = value.slice(0, -1);
+				} else {
+					log.warn(
+						`Missing semicolon in redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} in "${label}" block`,
+					);
+				}
+			}
+
+			if (key && value) {
+				parsedBlock[key] = value;
+			} else {
+				// Skip malformed lines
+				log.warn(
+					`Ignoring malformed redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} in "${label}" block due to missing key, value, or "="`,
+				);
+				continue;
+			}
+		}
+
+		return parsedBlock;
+	}
+}
+
+export async function readProxy(): Promise<HostProxyConfig | undefined> {
+	// Get and parse redsocks.conf
+	let rawConf: string | undefined;
+	try {
+		rawConf = await readFromBoot(redsocksConfPath, 'utf-8');
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return undefined;
+	}
+	const redsocksConf = RedsocksConf.parse(rawConf);
+
+	// Get and parse no_proxy
+	const noProxy = await readNoProxy();
+
+	// Build proxy object
+	const proxy = {
+		...HostConfig.fromRedsocksConfig(redsocksConf),
```

Comment:
I removed `toRedsocksConfig` and `fromRedsocksConfig` transformation methods after reviewing the commits again, since the patchBody.network.proxy is an object that is equal to `ProxyConfig` type after decoding to `HostConfiguration` (and loosely equal after decoding from `LegacyHostConfiguration`